src/operator/subgraph/partition_graph.cc

```diff
@@ -616,6 +619,83 @@ void CutGraphInputs(const std::vector<nnvm::NodeEntry*> &input_entries,
   }
 }
 
+/*!
+ * \brief Infer attrs for subgraph, given input nodes of subgraph from original graph
+ */
+nnvm::Graph InferSubgraphAttrs(
+    Graph* g, const std::vector<nnvm::NodeEntry>& orig_input_entries,
+    const std::unordered_map<const nnvm::Node*, nnvm::Symbol>& subgraphs,
+    nnvm::Graph&& sg) {
+  // return if partition without attrs
+  if (!g->HasAttr("context")) return std::move(sg);
+  const auto &idx_og = g->indexed_graph();
+  const auto &idx_g = sg.indexed_graph();
+  CHECK_EQ(idx_g.input_nodes().size(), orig_input_entries.size());
+
+  auto num_nodes = idx_g.num_node_entries();
+
+  auto orig_ctx = g->GetAttr<exec::ContextVector>("context");
+  auto orig_dev_masks = g->GetAttr<exec::DevMaskVector>("dev_mask");
+
+  auto oshapes = g->GetAttr<nnvm::ShapeVector>("shape");
+  auto odtypes = g->GetAttr<nnvm::DTypeVector>("dtype");
+  auto ostypes = g->GetAttr<mxnet::StorageTypeVector>("storage_type");
+
+  exec::ContextVector contexts(idx_g.num_nodes(), orig_ctx[0]);
+  nnvm::ShapeVector shapes(num_nodes);
+  nnvm::DTypeVector types(num_nodes, -1);
+  StorageTypeVector stypes(num_nodes, kUndefinedStorage);
+  exec::DevMaskVector dev_masks(idx_g.num_nodes(), orig_ctx[0].dev_mask());
+
+  nnvm::DFSVisit(sg.outputs, [&](const nnvm::NodePtr node) {
+    if (idx_og.exist(node.get())) {
+      auto nid = idx_g.node_id(node.get());
+      auto onid = idx_og.node_id(node.get());
+      contexts[nid] = orig_ctx[onid];
+      dev_masks[nid] = orig_dev_masks[onid];
+    }
+  });
+
+  const auto &input_nids = idx_g.input_nodes();
+  for (size_t i = 0; i < input_nids.size(); i++) {
+    auto nid = input_nids[i];
+    auto eid = idx_g.entry_id(input_nids[i], 0);
+    uint32_t onid = 0;
+    uint32_t oeid = 0;
+    if (idx_og.exist(orig_input_entries[i].node.get())) {
+      onid = idx_og.node_id(orig_input_entries[i].node.get());
+      oeid = idx_og.entry_id(orig_input_entries[i]);
+    } else {
+      auto previous = subgraphs.at(orig_input_entries[i].node.get());
+      onid = idx_og.node_id(previous.outputs[orig_input_entries[i].index].node.get());
+      oeid = idx_og.entry_id(previous.outputs[orig_input_entries[i].index]);
+    }
+
+    contexts[nid] = orig_ctx[onid];
+    dev_masks[nid] = orig_dev_masks[onid];
+
+    shapes[eid] = oshapes[oeid];
+    types[eid] = odtypes[oeid];
+    stypes[eid] = ostypes[oeid];
+  }
```

Comment:
could you please add comments about what lines 659-680 do?

Reply:
sure, will do.